src/textual/_styles_cache.py

```diff
@@ -0,0 +1,333 @@
+from __future__ import annotations
+
+import sys
+from typing import TYPE_CHECKING, Callable, Iterable, List
+
+from rich.segment import Segment
+from rich.style import Style
+
+from ._border import get_box, render_row
+from ._segment_tools import line_crop, line_pad, line_trim
+from ._types import Lines
+from .color import Color
+from .geometry import Region, Size, Spacing
+from .renderables.opacity import Opacity
+from .renderables.tint import Tint
+
+if sys.version_info >= (3, 10):
+    from typing import TypeAlias
+else:  # pragma: no cover
+    from typing_extensions import TypeAlias
+
+
+if TYPE_CHECKING:
+    from .css.styles import StylesBase
+    from .widget import Widget
+
+
+RenderLineCallback: TypeAlias = Callable[[int], List[Segment]]
+
+
+class StylesCache:
+    """Responsible for rendering CSS Styles and keeping a cached of rendered lines.
+
+    ```
+    ┏━━━━━━━━━━━━━━━━━━━━━━┓◀── A. border
+    ┃                      ┃◀┐
+    ┃                      ┃ └─ B. border + padding +
+    ┃   Lorem ipsum dolor  ┃◀┐         border
+    ┃   sit amet,          ┃ │
+    ┃   consectetur        ┃ └─ C. border + padding +
+    ┃   adipiscing elit,   ┃     content + padding +
+    ┃   sed do eiusmod     ┃           border
+    ┃   tempor incididunt  ┃
+    ┃                      ┃
+    ┃                      ┃
+    ┗━━━━━━━━━━━━━━━━━━━━━━┛
```

Comment:
Not following this diagram at all 😄

Reply:
Yeah saw those comments and understood them about the 3 code paths. Just didn't really get what the diagram was showing - wasn't getting that the arrows were meant to be point at a line/horizontal slice through the content box. I get it now though.

Reply:
I'll add a bit more text to explain, for future explorers of the code!